### ISSUE TITLE
"unbalanced" square brackets in subject break fetch

### ISSUE THREAD

**Issue by macdanny** [emersion/go-imap#212]
I have an email with a subject "l[[l". It is breaking go-imap, because it believes that these square brackets denote a response code and it treats the subject, plus everything that follows it, as a single atom when parsing.

This is the relevant debug output when fetching this message:

    * 2030 FETCH (FLAGS (\Seen nonjunk) INTERNALDATE "23-Dec-2006 19:55:36 -0700" RFC822.SIZE 1063 ENVELOPE ("Sat, 23 Dec 2006 19:55:36 -0700" l[[l ((DeepTech NIL deeptech147 hotpop.com)) ((DeepTech NIL deeptech147 hotpop.com)) ((DeepTech NIL deeptech147 hotpop.com)) ((NIL NIL freebsd-test freebsd.org)) () () "" <458DE5C1.3010802@hotpop.com>) UID 2030 BODY[] {1063}

I set breakpoints and stepped through the code ... it thinks the atom is "l[[l ((DeepTech...." all the way to the end. Eventually it fails because "list doesn't end with a close parenthesis". The list DOES end with a close parenthesis, the issue is that it thinks all of the parentheses on the line after the subject are part of the subject.

I have made a modification to the read code that allows the message o parse. I am concerned though that my change may not be valid. It seems like it should have been written this way to begin with. The fact that it wasn't makes me think that there's something else that depends on the functionality as it was.

My change is in the ReadAtom method. The old code:

	if char == respCodeEnd {
		if r.brackets == 0 {
			if r.inRespCode {
				break
			} else {
				return nil, newParseError("atom contains bad brackets nesting")
			}
		}
		r.brackets--
	}
	if char == respCodeStart {
		r.brackets++
	}

As I changed it ... essentially, I made ReadAtom treat brackets as normal characters unless r.inRespCode is true.

	if r.inRespCode {
		if char == respCodeEnd {
			if r.brackets == 0 {
				if r.inRespCode {
					break
				} else {
					return nil, newParseError("atom contains bad brackets nesting")
				}
			}
			r.brackets--
		}
		if char == respCodeStart {
			r.brackets++
		}
	}

This may even be a problem with the server, maybe it is required to quote the subject in this case. I have spent some time poring over the RFC and it doesn't look like it from my reading. I confess however I find the ABNF pretty cryptic.

**Comment by macdanny** [comment]
Actually I think it's the server that is non-conforming. I have the same email on two different IMAP servers. One of the servers quotes the subject and go-imap can read and parse it just fine. The other server doesn't quote the subject as I describe above. The ABNF mandates that the subject either be quoted or in the {7}literal form.

**Comment by emersion** [comment]
From the ABNF:

```
envelope        = "(" env-date SP env-subject SP env-from SP
                  env-sender SP env-reply-to SP env-to SP env-cc SP
                  env-bcc SP env-in-reply-to SP env-message-id ")"

env-subject     = nstring

nstring         = string / nil

string          = quoted / literal
```

So yes, this server seems to be non-conforming.

**Comment by macdanny** [comment]
I'll document my use case and where I'm at with it in case someone else runs into it, as unlikely as that might seem.

I'm migrating email from my own server onto ProtonMail through their ProtonMail Bridge application. That is the non-conforming IMAP server implementation. I have reported the bug to ProtonMail and they may fix it.

In the meantime, I further modified the ReadAtom method, adding additional code and setting some different breakpoints to detect legitimate uses of '[' and ']' characters that would be broken by the change I described above. I'm fetching all of the email (a few tens of thousands) I have previously migrated. So far, the only thing the change would functionally break is the ability to detect mismatched brackets on the BODY[] atom.

I'm not proposing that go-imap be changed, I'll vendor this library and keep my change local. It will be good enough for my needs. I am just documenting my findings.

**Comment by emersion** [comment]
Amusingly, their bridge uses go-imap. So go-imap's server is nonconforming.

**Comment by foxcpp** [comment]
Might be fixed by #261. Can somebody check it?

**Comment by emersion** [comment]
Yeah, it should fix it. Please re-open if you've updated to the latest version and still seeing this issue.
